serialization_v2/serialization.py

```diff
@@ -0,0 +1,1284 @@
+
+See :func:`serialize_config` for the main entry point.
+
+Note: Sisyphus hashes are currently just defined by the config keys/values,
+using the `sis_hash_helper` function, without any special handling,
+except for dim tags (RETURNN :class:`Dim` objects).
+That means, e.g. functions/classes get hashed by ``(obj.__module__, obj.__qualname__)``.
+We currently don't provide a way to customize the hashing behavior
+(except of ``post_config`` which is not hashed at all).
+This could be extended in the future by allowing more custom behavior
+e.g. for module scopes.
+Also, e.g. to specify ``unhashed_package_root`` for some of the references.
+
+Note: Sisyphus Path objects are serialized directly using :func:`sisyphus.Path.get_path`.
+
+We handle those objects specially:
+- primitive types (int, float, bool, str)
+- Sisyphus Path objects
+- RETURNN Dim objects
+- dict, list, tuple, set
+- functions, classes, modules
+- functools.partial (just some nicer repr)
+- i6_core.serialization objects
+- CodeWrapper
+
+All other generic objects are handled in the same way as pickle does it
+(or also :class:`i6_experiments.common.utils.dump_py_code.PythonCodeDumper`),
+i.e. using ``__reduce__`` (etc).
+This also allows circular references.
+"""
+
+from __future__ import annotations
+
+
+__all__ = ["serialize_config", "SisPathHandling", "ReturnnConfigWithNewSerialization", "SerializedConfig"]
+
+import base64
+import builtins
+from dataclasses import dataclass
+import enum
+import functools
+import inspect
+from keyword import iskeyword
+import math
+import os
+import pickle
+import re
+import subprocess
+import sys
+import types
+from types import BuiltinFunctionType, FunctionType, MethodType, ModuleType
+from typing import Any, Collection, Dict, List, Optional, Sequence, Tuple, Union, TYPE_CHECKING
+
+from sisyphus import tk, Path
+from sisyphus.delayed_ops import DelayedBase
+from sisyphus.hash import sis_hash_helper
+
+from i6_core.returnn.config import CodeWrapper, ReturnnConfig
+from i6_core.returnn.training import Checkpoint, PtCheckpoint
+from i6_core.serialization import (
+    Call,
+    CallImport,
+    CodeFromFile,
+    CodeFromFunction,
+    ExplicitHash,
+    ExternalImport,
+    Import,
+    NonhashedCode,
+    PartialImport,
+    SerializerObject,
+)
+from i6_core.serialization import Collection as SerializerCollection
+
+if TYPE_CHECKING:
+    from returnn.tensor import Dim
+
+
+def serialize_config(
+    config: Dict[str, Any],
+    post_config: Optional[Dict[str, Any]] = None,
+    *,
+    inlining: bool = True,
+    known_modules: Collection[str] = (),
+    extra_sys_paths: Sequence[str] = (),
+    returnn_root: Optional[Union[str, Path]] = None,
+    sis_path_handling: SisPathHandling = None,
+) -> SerializedConfig:
+    """serialize config. see module docstring for more info."""
+    if returnn_root is not None:
+        assert tk.running_in_worker(), "returnn_root only supported in sisyphus worker context"
```

Comment:
Why? While we currently mostly would use it that way, I don't see that this is really a restriction that is necessary.